LTR documentation (#8336)
opensearch-project/documentation-website: _search-plugins/ltr/logging-features.md

## `sltr` query

The `sltr` query is the primary method for running features and evaluating models. When logging, an `sltr` query is used to execute each feature query and retrieve the feature scores. A feature set structure is shown in the following example request:

```json
PUT _ltr/_featureset/more_movie_features
{
    "name": "more_movie_features",
    "features": [
        {
            "name": "body_query",
            "params": [
                "keywords"
                ],
            "template": {
                "match": {
                    "overview": "{% raw %}{{keywords}}{% endraw %}"
                }
            }
        },
        {
            "name": "title_query",
            "params": [
                "keywords"
            ],
            "template": {
                "match": {
                    "title": "{% raw %}{{keywords}}{% endraw %}"
                }
            }
        }
    ]
}
```
{% include copy-curl.html %}

## Common use cases

Common use cases for logging feature sets are described in the following sections.

### Joining feature values with a judgment list

> If the judgment list is already available, you can join feature values for each keyword/document pair to create a complete training set. For example, you have been provided the following judgment list:

Reviewer: ```suggestion
If the judgment list is already available, you can join feature values for each keyword/document pair to create a complete training set. For example, consider the following judgment list:
```

{: .note}    

To avoid influencing the score, the `sltr` query is injected as a filter, as shown in the following example:

```json
{
    "query": {
        "bool": {
            "filter": [
                {
                    "terms": {
                        "_id": [
                            "7555",
                            "1370",
                            "1369"
                        ]
                    }
                },
                {
                    "sltr": {
                        "_name": "logged_featureset",
                        "featureset": "more_movie_features",
                        "params": {
                            "keywords": "rambo"
                        }
                    }
                }
            ]
        }
    }
}
```
{% include copy-curl.html %}

Executing this query returns the three expected hits. The next step is to enable feature logging to refer to the `sltr` query to be logged.

> The logging identifies the `sltr` query, runs the feature set's queries, scores each document, and returns those scores as computed fields on each document, as shown in the following example logging:

Reviewer: ```suggestion
The logging identifies the `sltr` query, runs the feature set's queries, scores each document, and returns those scores as computed fields for each document, as shown in the following example logging structure:
```

> 

Reviewer: `logging operation`

```json
"ext": {
    "ltr_log": {
        "log_specs": {
            "name": "log_entry1",
            "named_query": "logged_featureset"
        }
    }
}
```
{% include copy-curl.html %}

The log extension supports the following arguments:

- `name`: The name of the log entry to fetch from each document.
- `named_query`: The named query that corresponds to an `sltr` query.
- `rescore_index`: If the `sltr` query is in a rescore phase, then this is the index of the query in the rescore list.
- `missing_as_zero`: Produces a `0` for missing features (when the feature does not match). Default is `false`.
  
To enable the logging to locate an `sltr` query, either for the normal query phase or during rescoring, either `named_query` or `rescore_index` must be set.

> {: .note}

Reviewer: `log`

> To enable the logging to locate an `sltr` query, either during the normal query phase or during rescoring, either `named_query` or `rescore_index` must be set.

Reviewer: ```suggestion
To enable the log to locate an `sltr` query, either during the normal query phase or during rescoring, either `named_query` or `rescore_index` must be set.
```

                        }
                }}
            ]
        }
    },
    "ext": {
        "ltr_log": {
            "log_specs": {
                "name": "log_entry1",
                "named_query": "logged_featureset"
            }
        }
    }
}
```
{% include copy-curl.html %}

Each document now contains a log entry, as shown in the following example:

```json
{
    "_index": "tmdb",
    "_type": "movie",
    "_id": "1370",
    "_score": 20.291,
    "_source": {
        ...
    },
    "fields": {
        "_ltrlog": [
            {
                "log_entry1": [
                    {"name": "title_query"
                     "value": 9.510193},
                    {"name": "body_query
                     "value": 10.7808075}
                ]
            }
        ]
    },
    "matched_queries": [
        "logged_featureset"
    ]
}
```
{% include copy-curl.html %}

The judgment list can be joined with the feature values to produce a training set. For the line corresponding to document `1370` with keyword `rambo`, the following can be added:

```
> 4 qid:1 1:9.510193 2:10.7808075
```
{% include copy-curl.html %}

Repeat this process for all of your queries.

> For large judgment lists, it is recommended to batch the logging for multiple queries. You can use [Multi search]({{site.url}}{{site.baseurl}}/api-reference/multi-search/) capabilities for this purpose.

Reviewer: ```suggestion
For large judgment lists, it is recommended to batch the logs for multiple queries. You can use [multi-search]({{site.url}}{{site.baseurl}}/api-reference/multi-search/) capabilities for this purpose.
```

{: .note}

### Logging values for a live feature set

If you are running in production with a model being executed within an `sltr` query, a live model may appear similar to the following example request:

```json
POST tmdb/_search
{
    "query": {
        "match": {
            "_all": "rambo"
        }
    },
    "rescore": {
        "query": {
            "rescore_query": {
                "sltr": {
                    "params": {
                        "keywords": "rambo"
                    },
                    "model": "my_model"
                }
            }
        }
    }
}
```
{% include copy-curl.html %}

See [Searching with LTR]({{site.url}}{{site.baseurl}}/search-plugins/ltr/searching-with-your-model/) for information about model execution.
{: .note}

> Applying the appropriate logging spec to reference the `sltr` query is the solution to log the feature values for the query, as shown in the following example: 

Reviewer: ```suggestion
To log the feature values for the query, apply the appropriate logging spec to reference the `sltr` query, as shown in the following example: 
```

```json
"ext": {
    "ltr_log": {
        "log_specs": {
            "name": "log_entry1",
            "rescore_index": 0
        }
    }
}
```
{% include copy-curl.html %}

> The example logs the features to the OpenSearch response, enabling future model retraining using the same feature set.

Reviewer: ```suggestion
The example logs the features in the response, enabling future model retraining using the same feature set.
```

                    "genres.name": "{% raw %}{{keywords}}{% endraw %}"
                }
            }
        }
    ]
}
```
{% include copy-curl.html %}

The feature set, `other_movie_features`, can be logged alongside the live production set, `more_movie_features`, by appending it as another filter, as shown in the following example request:

```json
POST tmdb/_search
{
"query": {
    "bool": {
        "filter": [
            { "sltr": {
                "_name": "logged_featureset",
                "featureset": "other_movie_features",
                "params": {
                    "keywords": "rambo"
                }
    }},
            {"match": {
                "_all": "rambo"
            }}
        ]
    }
},
"rescore": {
    "query": {
        "rescore_query": {
            "sltr": {
                "params": {
                    "keywords": "rambo"
                },
                "model": "my_model"
            }
        }
    }
}
}
```
{% include copy-curl.html %}

You can continue adding as many feature sets as needed for logging.

## Logging scenarios

Once you have covered the basics, you can consider some real-life feature logging scenarios.

> First, logging is used to develop judgment lists from user analytics to capture the exact value of a feature at the precise time of interaction. For instance, you may want to know the recency, title score, and other values at the precise time of a user's interaction. This would help you analyze what correlated with relevance while training. To achieve this, you can build a comprehensive feature set for future experimentation. 

Reviewer: ```suggestion
First, logging is used to develop judgment lists from user analytics to capture the exact value of a feature at the precise time of interaction. For instance, you may want to know the recency, title score, and other values at the precise time of a user's interaction. This would help you analyze which features or factors had relevance while training. To achieve this, you can build a comprehensive feature set for future experimentation. 
```

> First, logging is used to develop judgment lists from user analytics to capture the exact value of a feature at the precise time of interaction. For instance, you may want to know the recency, title score, and other values at the precise time of a user's interaction. This would help you analyze what correlated with relevance while training. To achieve this, you can build a comprehensive feature set for future experimentation. 

Reviewer: `which features or factors had`

---
layout: default
title: Logging feature scores
nav_order: 50
parent: Learning to Rank
has_children: false
---

# Logging feature scores

> Feature values need to be logged to train a model. This is a crucial component of the Learning to Rank plugin: as you search, feature values from the feature sets are logged so they can be used for training. This allows models that effectively predict relevance using that set of features to be discovered.

Reviewer: ```suggestion
Feature values need to be logged to train a model. This is a crucial component of the Learning to Rank plugin---as you search, feature values from the feature sets are logged so they can be used for training. This allows models that effectively predict relevance using that set of features to be discovered.
```

                "match": {
                    "genres.name": "{% raw %}{{keywords}}{% endraw %}"
                }
            }
        }
    ]
}
```
{% include copy-curl.html %}

The `other_movie_features` can be logged alongside the live production `more_movie_features` by appending it as another filter, as shown in the following example request:

```json
POST tmdb/_search
{
"query": {
    "bool": {
        "filter": [
            { "sltr": {
                "_name": "logged_featureset",
                "featureset": "other_movie_features",
                "params": {
                    "keywords": "rambo"
                }
    }},
            {"match": {
                "_all": "rambo"
            }}
        ]
    }
},
"rescore": {
    "query": {
        "rescore_query": {
            "sltr": {
                "params": {
                    "keywords": "rambo"
                },
                "model": "my_model"
            }
        }
    }
}
}
```
{% include copy-curl.html %}

You can continue adding as many feature sets as needed for logging.

## Purposes of logging

Once you have covered the basics, you might consider some real-life feature logging scenarios.

> First, you might develop judgment lists from user analytics. The goal is to capture the exact value of a feature at the precise time a user interaction occurred.If they clicked, you want to know the recency, title score, and every other value at that exact moment. This allows you can study later what correlated with relevance when training. To achieve this, you may build a large comprehensive feature set for future experimentation. 

Reviewer 2: This allows you **to** study later?

---
layout: default
title: Logging Feature Scores
nav_order: 50
parent: LTR search
has_children: false
---

# Logging Feature Scores

To train a model, you need to log feature values. This is a major
component of the learning to rank plugin: as users search, we log
feature values from our feature sets so we can then train. Then we can
discover models that work well to predict relevance with that set of
features.

## Sltr Query

The `sltr` query is the primary way features are run and models are
evaluated. When logging, we'll just use an `sltr` query for executing
every feature-query to retrieve the scores of features.

For the sake of discussing logging, let's say we created a feature set
like so that works with the TMDB data set from the
[demo](http://github.com/opensearch-project/opensearch-learning-to-rank-base/tree/main/demo/):

> 

Author: thanks...    There is a ticket out there to add this back into the repo...
